pkg/reconciler/tektonwrapper/tektonwrapper.go

```diff
@@ -0,0 +1,271 @@
+				Name:      twId,
+			}
+			if err := r.client.Get(ctx, twKey, &tw); err != nil {
+				return reconcile.Result{}, err
+			}
+			tw.Status.State = v1alpha1.TektonWrapperStateComplete
+			return reconcile.Result{}, r.client.Status().Update(ctx, &tw)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	switch tw.Status.State {
+	case v1alpha1.TektonWrapperStateComplete:
+		//TODO could initiate pruning of successful PR
+		return reconcile.Result{}, nil
+	case v1alpha1.TektonWrapperStateAbandoned:
+		//TODO could bump future metrics and initiate alerts to help notify admins and users that their projects are
+		// under powered
+		return reconcile.Result{}, nil
+	case v1alpha1.TektonWrapperStateInProgress:
+		//TODO even though we have the PipelineRun watch, we could list on our label and make sure the PipelineRun is
+		// active and we have not missed any events; for now, deferring on this
+		return reconcile.Result{}, nil
+	}
+
+	// get embedded PR
+	pr = &v1beta1.PipelineRun{}
+	if len(tw.Spec.PipelineRun) > 0 {
+		decodingScheme := runtime.NewScheme()
+		utilruntime.Must(v1beta1.AddToScheme(decodingScheme))
+		decoderCodecFactory := serializer.NewCodecFactory(decodingScheme)
+		decoder := decoderCodecFactory.UniversalDecoder(v1beta1.SchemeGroupVersion)
+		prHydrateErr := runtime.DecodeInto(decoder, tw.Spec.PipelineRun, pr)
+		if prHydrateErr != nil {
+			return reconcile.Result{}, prHydrateErr
+		}
+	} else {
+		msg := fmt.Sprintf("TektonWrapper %s:%s has no embedded PipelineRun", tw.Namespace, tw.Name)
+		log.Info(msg)
+		r.eventRecorder.Event(&tw, corev1.EventTypeWarning, "MissingPipelineRun", msg)
+		return reconcile.Result{}, nil
+	}
+	if pr == nil {
+		msg := fmt.Sprintf("TektonWrapper %s:%s could not produced a pipeline run with buf %#v", tw.Namespace, tw.Name, tw.Spec.PipelineRun)
+		log.Info(msg)
+		r.eventRecorder.Event(&tw, corev1.EventTypeWarning, "MissingPipelineRun", msg)
+		return reconcile.Result{}, nil
+	}
+
+	// we hydrate PipelineRun first to use it's namespace/name in an messaging
+	// see if we have quota for this namespace/project
+
+	// process quota
+	quotaList := quotav1.ClusterResourceQuotaList{}
+	//TODO controller runtime seemed unable to deal with openshift API and its attempt at mapping to CRDs; so for now
+	// we are using a non caching client; as such, we may not be able to afford to this list on every pass through this
+	// reconciler, but we need to get a sense for how often these quotas get defined in appstudio, when they are defined, etc.
+	qerr := r.nonCachingClient.List(ctx, &quotaList)
+	if qerr != nil {
+		return ctrl.Result{}, qerr
+	}
+	for _, quota := range quotaList.Items {
+		// find applicable quota for this namespace
+		ns, ok := quota.Spec.Selector.AnnotationSelector["openshift.io/requester"]
+
+		if !ok {
+			continue
+		}
+		if ns != request.Namespace {
+			continue
+		}
+		hardPodCount := 0
+		if quota.Spec.Quota.Hard.Pods() != nil {
+			hardPodCount = int(quota.Spec.Quota.Hard.Pods().Value())
```

Comment:
This could end up being an issue if we ever have pipelines with concurrent TaskRuns.

Reply:
good point ... I'll put a TODO comment for now; this approach still may just be a precursor to a more generic appstudio solution or pivot to the upstream tekton feature if that ever arrives 

worst case, if those generic appstudio/tekton do not land soon enough, and some form of this becomes our new norm, we'll deal with concurrent TaskRuns when we start using them.